### ISSUE TITLE
unable to install the package

### ISSUE THREAD

**Issue by akshaysiit** [Ensembles/ert#980]
I am trying to install this package on my macbook osx 10.11.2 and am getting this error - 

"Undefined symbols for architecture x86_64:
  "_dgemm_", referenced from:
      _matrix_dgemm in matrix_blas.c.o
  "_dgemv_", referenced from:
      _matrix_dgemv in matrix_blas.c.o
      _matrix_mul_vector in matrix_blas.c.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/libert_util.2.0.dylib] Error 1
make[1]: *** [libert_util/src/CMakeFiles/ert_util.dir/all] Error 2
make: *** [all] Error 2"

I am a novice when it comes to programming - can someone help me to understand this issue and how to solve it.

**Comment by jepebe** [comment]
It looks like your are missing *BLAS* and/or *LAPACK*. Be sure that these libraries are available to the linker while building *ERT*.

```BLAS_LIBRARY``` and ```LAPACK_LIBRARY``` in *cmake* must point to the correct *.dynlib*

If you do not have these libraries you can install them via **Homebrew** or **conda**

**Comment by akshaysiit** [comment]
I thought that those 2 libraries were already a part of OSX 10.11.2 - anyways I installed them.

But now I get a different error - 

"Undefined symbols for architecture x86_64:
  "__gfortran_concat_string", referenced from:
      _dgesvd_ in liblapack.a(dgesvd.o)
      _dormbr_ in liblapack.a(dormbr.o)
      _dormtr_ in liblapack.a(dormtr.o)
      _dtrtri_ in liblapack.a(dtrtri.o)
      _dormlq_ in liblapack.a(dormlq.o)
      _dormql_ in liblapack.a(dormql.o)
      _dormqr_ in liblapack.a(dormqr.o)
      ...
  "__gfortran_st_write", referenced from:
      _xerbla_ in liblapack.a(xerbla.o)
  "__gfortran_st_write_done", referenced from:
      _xerbla_ in liblapack.a(xerbla.o)
  "__gfortran_stop_string", referenced from:
      _xerbla_ in liblapack.a(xerbla.o)
  "__gfortran_string_len_trim", referenced from:
      _xerbla_ in liblapack.a(xerbla.o)
  "__gfortran_transfer_character_write", referenced from:
      _xerbla_ in liblapack.a(xerbla.o)
  "__gfortran_transfer_integer_write", referenced from:
      _xerbla_ in liblapack.a(xerbla.o)
  "_daxpy_", referenced from:
      _dstein_ in liblapack.a(dstein.o)
      _dlatrd_ in liblapack.a(dlatrd.o)
      _dsytd2_ in liblapack.a(dsytd2.o)
  "_dcopy_", referenced from:
      _dsyevx_ in liblapack.a(dsyevx.o)
      _dlarfb_ in liblapack.a(dlarfb.o)
      _dstein_ in liblapack.a(dstein.o)
      _dlasq1_ in liblapack.a(dlasq1.o)
  "_ddot_", referenced from:
      _dstein_ in liblapack.a(dstein.o)
      _dlatrd_ in liblapack.a(dlatrd.o)
      _dsytd2_ in liblapack.a(dsytd2.o)
  "_dgemm_", referenced from:
      _matrix_dgemm in matrix_blas.c.o
      _dgesvd_ in liblapack.a(dgesvd.o)
      _dgetrf_ in liblapack.a(dgetrf.o)
      _dgetri_ in liblapack.a(dgetri.o)
      _dgebrd_ in liblapack.a(dgebrd.o)
      _dgetrf2_ in liblapack.a(dgetrf2.o)
      _dlarfb_ in liblapack.a(dlarfb.o)
      ...
  "_dgemv_", referenced from:
      _matrix_dgemv in matrix_blas.c.o
      _matrix_mul_vector in matrix_blas.c.o
      _dgetri_ in liblapack.a(dgetri.o)
      _dlarft_ in liblapack.a(dlarft.o)
      _dlabrd_ in liblapack.a(dlabrd.o)
      _dlarf_ in liblapack.a(dlarf.o)
      _dlatrd_ in liblapack.a(dlatrd.o)
      ...
  "_dger_", referenced from:
      _dlarf_ in liblapack.a(dlarf.o)
  "_dnrm2_", referenced from:
      _dstein_ in liblapack.a(dstein.o)
      _dlarfg_ in liblapack.a(dlarfg.o)
  "_drot_", referenced from:
      _dbdsqr_ in liblapack.a(dbdsqr.o)
  "_dscal_", referenced from:
      _dsyevx_ in liblapack.a(dsyevx.o)
      _dbdsqr_ in liblapack.a(dbdsqr.o)
      _dgetrf2_ in liblapack.a(dgetrf2.o)
      _dorg2r_ in liblapack.a(dorg2r.o)
      _dstein_ in liblapack.a(dstein.o)
      _dlabrd_ in liblapack.a(dlabrd.o)
      _dlarfg_ in liblapack.a(dlarfg.o)
      ...
  "_dswap_", referenced from:
      _dgetri_ in liblapack.a(dgetri.o)
      _dsyevx_ in liblapack.a(dsyevx.o)
      _dbdsqr_ in liblapack.a(dbdsqr.o)
      _dsteqr_ in liblapack.a(dsteqr.o)
  "_dsymv_", referenced from:
      _dlatrd_ in liblapack.a(dlatrd.o)
      _dsytd2_ in liblapack.a(dsytd2.o)
  "_dsyr2_", referenced from:
      _dsytd2_ in liblapack.a(dsytd2.o)
  "_dsyr2k_", referenced from:
      _dsytrd_ in liblapack.a(dsytrd.o)
  "_dtrmm_", referenced from:
      _dlarfb_ in liblapack.a(dlarfb.o)
      _dtrtri_ in liblapack.a(dtrtri.o)
  "_dtrmv_", referenced from:
      _dlarft_ in liblapack.a(dlarft.o)
      _dtrti2_ in liblapack.a(dtrti2.o)
  "_dtrsm_", referenced from:
      _dgetrf_ in liblapack.a(dgetrf.o)
      _dgetri_ in liblapack.a(dgetri.o)
      _dgetrf2_ in liblapack.a(dgetrf2.o)
      _dgetrs_ in liblapack.a(dgetrs.o)
      _dtrtri_ in liblapack.a(dtrtri.o)
  "_idamax_", referenced from:
      _dgetrf2_ in liblapack.a(dgetrf2.o)
      _dstein_ in liblapack.a(dstein.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/libert_util.2.0.dylib] Error 1
make[1]: *** [libert_util/src/CMakeFiles/ert_util.dir/all] Error 2
make: *** [all] Error 2"

Any idea how to solve this ?

**Comment by jepebe** [comment]
How did you install the libraries? It seems that *gfortran* is missing. This library is usually a part of *gcc* and *clang*. 

I agree that LAPACK and BLAS is part of OSX and I found them in ```/System/Library/Accelerate.framework/Frameworks/vecLib.framework/Versions/A```

**Comment by GitPaean** [comment]
Old impression. OSX usually does not have fortran compilers by default, while have not used Mac for several years.

**Comment by akshaysiit** [comment]
for this purpose I did install gfortran 5.2 again before installing the libraries again.

And then tried to install ert again - but got the above error - any idea how to solve it now.

**Comment by jepebe** [comment]
The linker must be able to find the **.dylib** associated with *gfortran*. The library should be in one of the paths that can be found in *DYLD_LIBRARY_PATH* if it is not otherwise available.

**Comment by akshaysiit** [comment]
So what should/can I do ?

**Comment by jepebe** [comment]
How did you install LAPCK and BLAS?

**Comment by akshaysiit** [comment]
Using the details mentioned here - https://pheiter.wordpress.com/2012/09/04/howto-installing-lapack-and-blas-on-mac-os/

**Comment by jepebe** [comment]
When pull request: 984 is merged you should be able to build without installing separate **LAPACK** or **BLAS**. We have added **BLAS** as a required link dependency during building. 

On my Mac running OSX 10.11 and XCode 7.2 I was able to build without any external dependencies (except for CMake).

**Comment by akshaysiit** [comment]
tried again after merging - now got this error

"Undefined symbols for architecture x86_64:
  "__gfortran_concat_string", referenced from:
      _dgesvd_ in liblapack.a(dgesvd.o)
      _dormbr_ in liblapack.a(dormbr.o)
      _dormtr_ in liblapack.a(dormtr.o)
      _dtrtri_ in liblapack.a(dtrtri.o)
      _dormlq_ in liblapack.a(dormlq.o)
      _dormql_ in liblapack.a(dormql.o)
      _dormqr_ in liblapack.a(dormqr.o)
      ...
  "__gfortran_st_write", referenced from:
      _xerbla_ in liblapack.a(xerbla.o)
  "__gfortran_st_write_done", referenced from:
      _xerbla_ in liblapack.a(xerbla.o)
  "__gfortran_stop_string", referenced from:
      _xerbla_ in liblapack.a(xerbla.o)
  "__gfortran_string_len_trim", referenced from:
      _xerbla_ in liblapack.a(xerbla.o)
  "__gfortran_transfer_character_write", referenced from:
      _xerbla_ in liblapack.a(xerbla.o)
  "__gfortran_transfer_integer_write", referenced from:
      _xerbla_ in liblapack.a(xerbla.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/libert_util.2.0.dylib] Error 1
make[1]: *** [libert_util/src/CMakeFiles/ert_util.dir/all] Error 2
make: *** [all] Error 2"

really confused now - all the time the error occurs when building this - "Linking C shared library ../../lib/libert_util.dylib"

**Comment by jepebe** [comment]
This is what I did without installing **LAPACK** or **BLAS**
```
git clone https://github.com/Ensembles/ert
cd ert
mkdir build
cd build
cmake ../devel/
make -j 4
```
Running  ```ccmake .``` in *build* I got among other values:
```
BLAS_LIBRARY      /usr/lib/libblas.dylib
LAPACK_LIBRARY    /usr/lib/liblapack.dylib
```
If you have these values and the **.dylib** files are the ones provided by OSX then hopefully it should work.

**Comment by akshaysiit** [comment]
I was doing the same except for the last line - I had it as "make ; make install" does that and this make any difference ?

**Comment by jepebe** [comment]
Not really. 
```make -j 4``` Means building in parallel with 4 processes.
```make install``` Installs the results on your system.

I fear that if you built and installed **LAPACK** and **BLAS** you may overwritten the original OSX libraries. I would check that first.

**Comment by akshaysiit** [comment]
How do I do that ?

**Comment by akshaysiit** [comment]
and how do i fix it ?

**Comment by jepebe** [comment]
I believe that you can find the original libraries in:
```/System/Library/Accelerate.framework/Frameworks/vecLib.framework/Versions/A```
and that the original libraries in **/usr/lib** were symlinks to the ones in there.

**Comment by akshaysiit** [comment]
I dont see any changes to them. Anyways is there a way to get back to original libraries before i built and installed LAPACK and BLAS

**Comment by jepebe** [comment]
if you do ``` ls -l /usr/lib ```

Then you should see something like this for **/usr/lib/liblapack.dylib**:

``` 
lilapack.dylib -> /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib 
```

And something similar for **/usr/lib/libcblas.dylib**. If this is not the case then that must be fixed.

**Comment by akshaysiit** [comment]
Yup, got both of them the same. So this means that those are ok, right ? If so what can I do to fix those errors above ?

**Comment by jepebe** [comment]
If **cmake** points to **/usr/lib/liblapack.dylib** and **/usr/lib/libcblas.dylib** then I do not know. The missing *__gfortran* are only a problem with the manually installed **LAPACK** and **BLAS** so I do not know why they are picked up. 

Did you completely remove the files you installed via this:  https://pheiter.wordpress.com/2012/09/04/howto-installing-lapack-and-blas-on-mac-os/ ?

**Comment by akshaysiit** [comment]
Nope did not remove those files ? how can i do that ?

**Comment by jepebe** [comment]
I do not know. I have only used **Homebrew** for installing libraries on OSX.

**Comment by pgdr** [comment]
I'm voting to close this issue since we haven't heard back from reporter.  We could probably improve some routines, cmake-output, guidelines etc., but that belong to a separate issue.

**Comment by joakim-hove** [comment]
> I'm voting to close this issue 

:+1:
